src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts

```diff
@@ -12,13 +67,205 @@ import { ActivatedRoute } from '@angular/router';
+  }
+
+  /**
+   * Retrieves user data rows in the spreadsheet interface and transforms it into a string.
+   *
+   * Null value from cell is changed to empty string after .join(). Filters empty rows in the process.
+   *
+   * Example:
+   * 2 by 5 spreadsheetData (before)
+   * ['TestSection1', 'Team1', 'null', 'test1@xample.com', 'test1comments']
+   * ['TestSection2', null, 'TestName2', 'test2@example.com', null]
+   *
+   * 2 by 5 spreadsheetData (after)
+   * "TestSection1|Team1||test1@xample.com|test1comments\n
+   *  TestSection2||TestName2|test2@example.com|\n"
+   */
+  getUserDataRows(spreadsheetData: string[][]): string {
+    // needs to check for '' as an initial empty row with null values will be converted to e.g. "||||" after .map
+    return spreadsheetData.filter((row: string[]) => (!row.every((cell: string) => cell === null || cell === '')))
+        .map((row: string[]) => row.join('|'))
+        .map((row: string) => row.replace(/\n|\r/g, ''))
+        .join('\n');
+  }
+
+  /**
+   * Submits enroll data
+   */
+  submitEnrollData(): void {
+    const newStudentsHOTInstance: Handsontable =
+        this.hotRegisterer.getInstance(this.newStudentsHOT);
+    const spreadsheetData: string[][] = newStudentsHOTInstance.getData();
+
+    const hotInstanceColHeaders: string[] = (newStudentsHOTInstance.getColHeader() as string[]);
+    const dataPushToTextarea: string =
+        this.getUpdatedHeaderString(hotInstanceColHeaders);
+    const userDataRows: string = this.getUserDataRows(spreadsheetData);
+
+    this.enrollData = (userDataRows === ''
+        ? '' : dataPushToTextarea + userDataRows); // only include header string if userDataRows is not empty
+
+    const paramMap: { [key: string]: string } = { courseid: this.courseid };
+    this.httpRequestService.post('/courses/enrollSave', paramMap, this.enrollData)
+        .subscribe((resp: EnrollResultPanelList) => {
+          this.showEnrollResults = true;
+          this.statusMessage.pop(); // removes any existing status message
+          this.statusMessage.push({
+            message: 'Enrollment Successful. Summary given below',
+            color: 'success',
+          });
+          this.enrollResultPanelList = resp.enrollResultPanelList;
+        }, (resp: ErrorMessageOutput) => {
+          this.statusMessage.pop(); // removes any existing status message
+          this.statusMessage.push({
+            message: resp.error.message,
+            color: 'danger',
+          });
+        });
+  }
+
+  /**
+   * Adds new rows to the 'New students' spreadsheet interface
+   * according to user input
+   */
+  addRows(numOfRows: number): void {
+    this.hotRegisterer.getInstance(this.newStudentsHOT).alter(
+        'insert_row', [], numOfRows);
+  }
+
+  /**
+   * Toggles the view of spreadsheet interface
+   * and/or its affiliated buttons
+   */
+  togglePanel(event: any): void {
+    this.targetElement =
+        event.target.closest('div').querySelector('i');
```

Comment:
Is there a way that's not dependent on DOM structure? Direct DOM manipulation is strongly discouraged in Angular.

Reply:
Let me figure something out using `ngClass`.